python/ray/experimental/state/api.py

```diff
@@ -1,195 +1,239 @@
+
+        Returns:
+            A list of queried result from `ListApiResponse`,
+
+        Raises:
+            This doesn't catch any exceptions raised when the underlying request
+            call raises exceptions. For example, it could raise `requests.Timeout`
+            when timeout occurs.
+
+        """
+        # Append the ListApiOptions to the url
+        endpoint = f"/api/v0/{resource.value}?{self._get_query_string(options)}"
+
+        # We don't use `asdict` to avoid deepcopy.
+        # https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict
+        params = {
+            "limit": options.limit,
+            "timeout": options.timeout,
+            "filter_keys": [],
+            "filter_values": [],
+        }
+        for filter in options.filters:
+            filter_k, filter_val = filter
+            params["filter_keys"].append(filter_k)
+            params["filter_values"].append(filter_val)
+
+        # NOTE(rickyyx)
+        # This might raise `requests.Timeout` exceptions and other related requests
+        # exceptions. Users of the functions are expected to handle this.
+        response = self._do_request(
+            "GET",
+            endpoint,
+            timeout=options.timeout,
+            params=params,
+        )
+
+        if response.status_code != 200:
+            self._raise_error(response)
+
+        response = response.json()
+        if response["result"] is False:
+            raise RayStateApiException(
+                "API server internal error. See dashboard.log file for more details. "
+                f"Error: {response['msg']}"
+            )
 
-    if _explain:
         # Print warnings if anything was given.
-        warning_msg = response["data"].get("partial_failure_warning", None)
-        if warning_msg:
-            warnings.warn(warning_msg, RuntimeWarning)
+        warning_msgs = response["data"].get("partial_failure_warning", None)
+        if warning_msgs and _explain:
+            warnings.warn(warning_msgs, RuntimeWarning)
+
+        return response["data"]["result"]
+
+
+"""
+Convenient methods for list_<RESOURCE>
 
-    return r.json()["data"]["result"]
+Supported arguments to the below methods, see `ListApiOptions`:
+    address: The address of the Ray state server. If None, it assumes a running Ray
+        deployment exists and will query the GCS for auto-configuration.
+    timeout: Time for the request.
+    limit: Limit of entries in the result
+"""
 
 
 def list_actors(
-    api_server_url: str = None,
+    address: Optional[str] = None,
     filters: List[Tuple[str, SupportedFilterType]] = None,
     limit: int = DEFAULT_LIMIT,
     timeout: int = DEFAULT_RPC_TIMEOUT,
     _explain: bool = False,
 ):
-    return _list(
-        "actors",
-        ListApiOptions(limit=limit, timeout=timeout, filters=filters),
-        api_server_url=api_server_url,
+    return StateApiClient(api_server_address=address).list(
```

Comment:
What's the overhead of initializing a new client? I think when this is used as a SDK, (python API), we don't want to create the client every time (unless initialization of the client has almost no cost).

Reply:
One potential solution is to require the client as an input value and expose the client to users. Looks like the job submission uses that approach. https://docs.ray.io/en/master/cluster/job-submission.html#python-sdk Another is we don't support these Python APIs but ask users to directly use the client.

Reply:
Good point! 
I do like the SDK style API, but I guess something like this is always pythony:
```
with StateApiClient(address) as client:
   client.list(...)
```

Or, we could support both? Because sometimes users would just do a single list (I don't really have insights in usecases though). They could choose to cache the client if they don't want to pay for the extra cost and prefer less verbose code. 
If this ^ I will document that in the header for usage.